File: tests/Examples/openfhe/ckks/matvec_512x784/matvec_512x784_test.cpp
#include <vector>

#include "gtest/gtest.h"  // from @googletest
#include "tests/Examples/openfhe/ckks/matvec_512x784/matvec_512x784_lib.h"

TEST(Matvec512x784Test, RunTest) {
  auto cryptoContext = matvec__generate_crypto_context();
  auto keyPair = cryptoContext->KeyGen();
  auto publicKey = keyPair.publicKey;
  auto secretKey = keyPair.secretKey;
  cryptoContext = matvec__configure_crypto_context(cryptoContext, secretKey);

  int cols = 784;
  int rows = 512;
  std::vector<float> arg0(cols, 0.1);

  float expected = 78.4;

  auto ct0 = matvec__encrypt__arg0(cryptoContext, arg0, publicKey);

  auto resultCt = matvec(cryptoContext, ct0);

  auto result = matvec__decrypt__result0(cryptoContext, resultCt, secretKey);

  float errorThreshold = 0.1;
  for (int i = 0; i < rows; i++) {
    EXPECT_NEAR(expected, result[i], errorThreshold);
  }
}

File: tests/Examples/lattigo/ckks/matvec_512x784/matvec_512x784_test.go
package matvec512x784

import (
	"math"
	"testing"
)

func TestMatvec(t *testing.T) {
	evaluator, params, ecd, enc, dec := Matvec__configure()

	cols := 784
	rows := 512
	arg0 := make([]float32, cols)
	for i := 0; i < cols; i++ {
		arg0[i] = 0.1
	}

	expected := float32(78.4)
	ct0 := Matvec__encrypt__arg0(evaluator, params, ecd, enc, arg0)
	resultCt := Matvec(evaluator, params, ecd, ct0)
	result := Matvec__decrypt__result0(evaluator, params, ecd, dec, resultCt)
	// Error threshold increased to 4.0 due to fallback to Halevi-Shoup kernel
	// which has different noise characteristics.
	errorThreshold := float64(4.0)
	for i := 0; i < rows; i++ {
		if math.Abs(float64(result[i]-expected)) > errorThreshold {
			t.Errorf("Decryption error at index %d: %.2f != %.2f", i, result[i], expected)
		}
	}
}
